feat: IStakingModuleV2 in CSM

madlabman opened the pull request:
## Description

Implement `IStakingModuleV2` interface for CSM.

## Checklist

- [x] Appropriate PR labels applied
- [x] Test coverage maintained (`just coverage`)
  - [x] Tests are added/updated
- [x] Documentation maintained
  - [x] No need to update

madlabman committed:
feat: IStakingModuleV2 in CSM

dgusakov commented on src/CSModule.sol:
```diff
@@ -53,9 +70,10 @@ contract CSModule is ICSModule, BaseModule {
-        onlyRole(STAKING_ROUTER_ROLE)
         returns (bytes memory publicKeys, bytes memory signatures)
     {
+        _checkRole(STAKING_ROUTER_ROLE);
```
But why?

madlabman replied:
To save space on the stack. Added a comment.

dgusakov commented on .github/workflows/test.yml:
```diff
@@ -125,7 +126,7 @@ jobs:
       - name: Build
         run: just build --sizes
-        # continue-on-error: true
+        continue-on-error: true # FIXME: Comment out after fixing the code size issues.
```
Are these lines and the change above for different purposes? If so, pls specify both

dgusakov commented on src/CSModule.sol:
```diff
@@ -176,6 +233,107 @@ contract CSModule is ICSModule, BaseModule {
+
+            if (depositAmount == 0 && keyTopUpLimit > 0) {
+                revert InvalidTopUpOrder();
+            }
```
Not sure we need to revert here. The condition basically means that we are out of depositAmount but there are still some keys to top up and it is fine. I think we should just return here. WDYT?

madlabman replied:
Please check the updated implementation.

dgusakov commented on src/CSModule.sol:
```diff
@@ -176,6 +233,107 @@ contract CSModule is ICSModule, BaseModule {
+            allocations[i] = Math.min(keyTopUpLimit, depositAmount);
+            depositAmount -= allocations[i];
+
+            if (allocations[i] == keyTopUpLimit) {
```
Haven't checked but it looks like saving current allocation to a scalar var might be cheaper than extracting array element

madlabman replied:
We're really tight on stack space here, that's why the reading from the memory.

dgusakov commented on src/CSModule.sol:
```diff
@@ -176,6 +233,107 @@ contract CSModule is ICSModule, BaseModule {
+    ) external onlyActiveTopUpQueue onlyRole(MANAGE_TOP_UP_QUEUE_ROLE) {
+        _topUpQueue().limit = limit.toUint32();
+        emit TopUpQueueLimitSet(limit);
+        _incrementModuleNonce();
```
Not sure if the key's state is changed when we modify the limit. I suggest not increasing nonce here

madlabman replied:
Depositable keys count may change, hence, the nonce change.

madlabman replied:
Depositable keys count may change, hence, the  nonce count change.

dgusakov commented on src/CSModule.sol:
```diff
@@ -327,8 +521,50 @@ contract CSModule is ICSModule, BaseModule {
+        }
+    }
+
+    function _initTopUpQueue(uint32 topUpQueueLimit) internal {
```
Worth adding a @dev comment that if the limit is zero on init than the queue is deactivated permanently

dgusakov commented on src/abstract/BaseModule.sol:
```diff
@@ -743,21 +743,6 @@ abstract contract BaseModule is
     }
 
-    /// @inheritdoc IStakingModule
-    function getStakingModuleSummary()
```
Keep the signature but make method virtual? Same for the other essential methods I guess

madlabman replied:
It becomes pretty ugly if you need to override it, because we have to specify both `IStakingModule` and `BaseModule` there.

dgusakov replied:
Not sure I got you

madlabman replied:
You would have to write something like this:
```solidity
function getStakingModuleSummary() overrides(IStakingModule, BaseModule) ...
```
At the same time it seems like all the modules use different implementation anyway.

dgusakov commented on src/lib/PackedPubkeys.sol:
```diff
@@ -0,0 +1,28 @@
+        uint256 keyIndex
+    ) internal pure returns (bytes memory key) {
+        key = new bytes(PUBKEY_LENGTH);
+
+        assembly ("memory-safe") {
+            let p := mul(PUBKEY_LENGTH, keyIndex)
+            if gt(add(p, PUBKEY_LENGTH), self.length) {
+                mstore(0x00, 0x4e487b71) // `Panic(uint256)`.
+                mstore(0x20, 0x32) // Out of bounds.
+                revert(0x1c, 0x24)
+            }
+
+            p := add(self.offset, p)
+            // evmcodes: for out of bound bytes, 0s will be copied.
+            calldatacopy(add(key, 0x20), p, PUBKEY_LENGTH)
```
If this is identical or similar to what we use in SigningKeysLib, is it worth using a single impl?

madlabman replied:
It would require refactoring of `SigningKeysLib` and agreeing to tradeoffs coming with it. The function is optimized for consuming one key at a time with boundary check, that makes sense for our usage.

dgusakov commented on src/lib/TopUpQueueLib.sol:
```diff
@@ -0,0 +1,102 @@
+
+function noId(TopUpQueueItem self) pure returns (uint32 n) {
+    assembly ("memory-safe") {
+        n := shr(32, self)
```
Add comments like "Take first 32 bytes". Here and after pls

dgusakov commented on src/lib/TopUpQueueLib.sol:
```diff
@@ -0,0 +1,102 @@
+        item = self.items[self.head];
+    }
+
+    function capacity(Queue storage self) internal view returns (uint256) {
```
`availableCapacity`?

madlabman replied:
I think `capacity` is good in conjunction with `limit`.

dgusakov replied:
Not sure. But I will leave it up to you

dgusakov reviewed (changes requested):
Haven't looked at tests for now

vgorkavenko commented on src/interfaces/IStakingModule.sol:
```diff
@@ -230,28 +230,30 @@ interface IStakingModule {
     /// @param keyIndices List of keys' indices
     /// @param operatorIds List of operator indices
-    /// @param topUpLimitsGwei List of gwei amount that can be deposited to key based on Cl data and SR logic
+    /// @param topUpLimits List of gwei amount that can be deposited to key based on Cl data and SR logic
```
Is it `topUpLimitsGwei` due to hackmd ?

madlabman replied:
Everything is in Gweis here, and it looks confusing to specify units in one place and not do so in another. I think I've pointed it out to the SR team already.

madlabman committed:
test: check allocations in top-up tests

madlabman committed:
feat: allocate 0 to keys when deposit amount is exhausted

madlabman committed:
refactor: TopUpQueueItem functions

madlabman committed:
chore: comment

madlabman committed:
chore: comment in workflow

dgusakov commented on src/CSModule.sol:
```diff
@@ -40,6 +62,23 @@ contract CSModule is ICSModule, BaseModule {
+        _initTopUpQueue(topUpQueueLimit);
+    }
+
+    /// @dev This method is expected to be called only when the contract is upgraded from version 2 to version 3 for the existing version 2 deployment.
+    ///      If the version 3 contract is deployed from scratch, the `initialize` method should be used instead.
+    function finalizeUpgradeV3(
+        uint32 topUpQueueLimit
+    ) external reinitializer(3) {
+        _initTopUpQueue(topUpQueueLimit);
+    }
```
Since we have no 0x02 modules to be upgraded, I suggest explicitly setting topUpQueue limit to 0 here

dgusakov commented on src/CSModule.sol:
```diff
@@ -119,14 +160,30 @@ contract CSModule is ICSModule, BaseModule {
+                            _topUpQueue().enqueue(
+                                newTopUpQueueItem(
+                                    // The ids are assigned sequentially, so noId can't exceed uint32 in practice.
+                                    // forge-lint: disable-next-line(unsafe-typecast)
```
Why not uint32() explicitly?

madlabman replied:
Removed the leftover comment.

dgusakov commented on src/CSModule.sol:
```diff
@@ -176,6 +233,105 @@ contract CSModule is ICSModule, BaseModule {
+            } else {
+                itemsSkipped++;
+            }
+        }
```
Why do we have skipped Items? I see the algo like this:
- Iterate over the queue
- Allocate min(topUpLimit, depositsAmountLeft)
- when depositAmountLeft == 0 or there are no more items in the queue - exit

madlabman replied:
We should not stop on `depositAmount == 0` to be able to process a queue where all or some keys has no top-up capacity. `itemsSkipped` was used to serve another case to return zero allocations for keys above deposit limit. Now I see we don't need this, so I changed the logic to revert if we provide keys that do not reflect the queue.

madlabman committed:
chore: disable top-up queue in `finalizeUpgradeV3`

madlabman committed:
chore: remove comment

madlabman committed:
refactor: revert when keys above deposit amount

madlabman committed:
Merge branch 'develop' into istaking-module-v2-csm

dgusakov commented on src/CSModule.sol:
```diff
@@ -176,6 +232,109 @@ contract CSModule is ICSModule, BaseModule {
+            if (allocations[i] == topUpLimits[i]) {
+                _topUpQueue().dequeue();
+            } else {
+                lastItemPartialDeposit = true;
```
We should accept cases when the keys passed to the function can accommodate more than depositAmount in total.

madlabman replied:
We do. But the function expects no more keys after full deposit amount utilization. We can return zero allocations for the rest of the keys, but it makes little to no sense in the current implementation. If we removed keys from the queue out of order when there's no more capacity left on them, it would make sense though.

dgusakov commented on src/CSModule.sol:
```diff
@@ -327,8 +522,51 @@ contract CSModule is ICSModule, BaseModule {
+    /// @dev Setting `topUpQueueLimit` to 0 effectively disables the top-up queue permanently.
+    function _initTopUpQueue(uint32 topUpQueueLimit) internal {
+        if (topUpQueueLimit > 0) {
+            _topUpQueue().limit = topUpQueueLimit;
```
Emit event?

dgusakov commented on src/lib/TopUpQueueLib.sol:
```diff
@@ -0,0 +1,93 @@
+        Queue storage self,
+        uint256 index
+    ) internal view returns (TopUpQueueItem) {
+        index = self.head + index;
```
This approach differs from our current deposit queue process. I suggest using th same approach in both queues

madlabman replied:
Do we need the ability to remove an item from the queue? If not, the current implementation is simpler and more storage‑efficient but admittedly is different.

dgusakov commented on src/CSModule.sol:
```diff
@@ -176,6 +232,109 @@ contract CSModule is ICSModule, BaseModule {
+        bool lastItemPartialDeposit;
+        for (uint256 i; i < keyIndices.length; i++) {
+            if (lastItemPartialDeposit) {
+                revert InvalidTopUpOrder();
```
TooManyKeysForTopUp

dgusakov replied:
Inform @mkurayan about this behaviour

dgusakov commented on src/lib/PackedPubkeys.sol:
```diff
@@ -0,0 +1,28 @@
+        assembly ("memory-safe") {
+            let p := mul(PUBKEY_LENGTH, keyIndex)
+            if gt(add(p, PUBKEY_LENGTH), self.length) {
+                mstore(0x00, 0x4e487b71) // `Panic(uint256)`.
```
```suggestion
                // Equals to revert Panic(0x32) where 0x32 stands for "Out of bounds" in human-readable format
                mstore(0x00, 0x4e487b71) // `Panic(uint256)`.
```

dgusakov commented on src/lib/TopUpQueueLib.sol:
```diff
@@ -0,0 +1,93 @@
+        }
+
+        // NOTE: Zeroing out the storage slot, since it's unreachable after `dequeue`.
+        self.items[self.head] = TopUpQueueItem.wrap(0x00);
```
Use a circular buffer with size 256

madlabman replied:
I've tried to implement one [here](https://github.com/lidofinance/community-staking-module/commit/c6067e21dc5913e69fc993881ee77bb62ceaf04a#diff-8ac9b76298fbab09dff3c79b52814c426ef7081630e9af61982fceabbcd2b9d9). So far I didn't like it much, so I suggest to return to it later.

madlabman committed:
refactor: revert with special error on extra item after partial top-up

madlabman committed:
chore: add note about Panic error in assembly

madlabman committed:
fix: emit TopUpQueueLimitSet in _initTopUpQueue

madlabman committed:
Merge remote-tracking branch 'origin/develop' into istaking-module-v2-csm

madlabman committed:
chore: disable code size limit on implementation deploy if needed

madlabman committed:
test: fix exited in SR integration tests

vgorkavenko commented on src/CSModule.sol:
```diff
@@ -191,6 +234,112 @@ contract CSModule is ICSModule, BaseModule {
+        uint256[] calldata topUpLimits
+    )
+        external
+        onlyActiveTopUpQueue
```
Does StakingRouter handle reverts correctly?

madlabman replied:
I think it's safe to assume it does, because this function validates keys and SR should fail if we revert here.

madlabman committed:
Bytecode squeeze (#658)

## Description

Extract function bodies to external libraries. Applying byte code saving
tricks here and there.

Affected functions:
- `getNodeOperatorSummary`
- `obtainDepositData` (top-ups)
- `cleanDepositQueue`
- `createNodeOperator`
- `reportWithdrawnValidators`
- bunch of functions using `_onOperatorDepositableChange`
- bunch of functions using `_setTargetLimit`

```patch
--- <unnamed>
+++ <unnamed>
@@ -186,17 +186,17 @@
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | changeNodeOperatorRewardAddress                                | 25947           | 32761  | 32747  | 40960   | 8       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| cleanDepositQueue                                              | 22412           | 74098  | 71792  | 111584  | 20      |
-|----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| compensateGeneralDelayedPenalty                                | 24510           | 59143  | 65548  | 90546   | 5       |
+| cleanDepositQueue                                              | 25784           | 68224  | 66146  | 107080  | 21      |
+|----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
+| compensateGeneralDelayedPenalty                                | 24510           | 56863  | 65506  | 84867   | 5       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | confirmNodeOperatorManagerAddressChange                        | 27694           | 30057  | 29840  | 33074   | 5       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | confirmNodeOperatorRewardAddressChange                         | 27281           | 30668  | 31026  | 32626   | 6       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| createNodeOperator                                             | 26454           | 116422 | 121932 | 144902  | 375     |
-|----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| decreaseVettedSigningKeysCount                                 | 26540           | 67686  | 79733  | 106702  | 27      |
+| createNodeOperator                                             | 26454           | 120200 | 125705 | 148827  | 376     |
+|----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
+| decreaseVettedSigningKeysCount                                 | 26540           | 63560  | 74051  | 99320   | 27      |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | depositQueueItem                                               | 0               | 662    | 0      | 2934    | 217     |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
@@ -210,7 +210,7 @@
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | getKeysForTopUp                                                | 4169            | 27557  | 28698  | 48378   | 6       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| getNodeOperator                                                | 0               | 6995   | 13469  | 13469   | 953     |
+| getNodeOperator                                                | 0               | 7002   | 13469  | 13469   | 954     |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | getNodeOperatorIds                                             | 3082            | 3437   | 3412   | 3949    | 8       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
@@ -222,7 +222,7 @@
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | getNodeOperatorOwner                                           | 6245            | 6245   | 6245   | 6246    | 2       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| getNodeOperatorSummary                                         | 30983           | 31374  | 31500  | 31924   | 36      |
+| getNodeOperatorSummary                                         | 32684           | 33076  | 33201  | 33629   | 36      |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | getNodeOperatorTotalDepositedKeys                              | 3866            | 3866   | 3866   | 3866    | 2       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
@@ -264,7 +264,7 @@
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | obtainDepositData(uint256,bytes)                               | 26220           | 116794 | 96728  | 274625  | 114     |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| obtainDepositData(uint256,bytes,uint256[],uint256[],uint256[]) | 26873           | 47207  | 47244  | 75161   | 16      |
+| obtainDepositData(uint256,bytes,uint256[],uint256[],uint256[]) | 26873           | 52558  | 53451  | 84438   | 16      |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | onExitedAndStuckValidatorsCountsUpdated                        | 23095           | 23095  | 23095  | 23095   | 1       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
@@ -286,13 +286,13 @@
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | recoverEther                                                   | 24287           | 26544  | 26544  | 28802   | 2       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| removeKeys                                                     | 25154           | 94354  | 104065 | 181682  | 18      |
-|----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| reportGeneralDelayedPenalty                                    | 25439           | 99195  | 112623 | 112936  | 29      |
+| removeKeys                                                     | 25154           | 92455  | 100883 | 179136  | 18      |
+|----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
+| reportGeneralDelayedPenalty                                    | 25439           | 96779  | 109444 | 109757  | 29      |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | reportValidatorExitDelay                                       | 28337           | 30179  | 30179  | 32022   | 2       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| reportWithdrawnValidators                                      | 25844           | 109084 | 113651 | 235932  | 44      |
+| reportWithdrawnValidators                                      | 25844           | 113459 | 120514 | 235720  | 44      |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | resetNodeOperatorManagerAddress                                | 27633           | 32639  | 30911  | 39180   | 6       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
@@ -302,17 +302,17 @@
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | setTopUpQueueLimit                                             | 23806           | 39722  | 36580  | 53680   | 9       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| settleGeneralDelayedPenalty                                    | 25056           | 52592  | 45659  | 89158   | 15      |
+| settleGeneralDelayedPenalty                                    | 25056           | 55266  | 45659  | 97305   | 15      |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | supportsInterface                                              | 363             | 480    | 480    | 598     | 2       |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | unsafeUpdateValidatorsCount                                    | 25443           | 36445  | 35872  | 45690   | 11      |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| updateDepositableValidatorsCount                               | 44202           | 59542  | 58737  | 104912  | 38      |
+| updateDepositableValidatorsCount                               | 44160           | 56883  | 55789  | 107924  | 38      |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
 | updateExitedValidatorsCount                                    | 26242           | 41270  | 47129  | 56792   | 11      |
 |----------------------------------------------------------------+-----------------+--------+--------+---------+---------|
-| updateTargetValidatorsLimits                                   | 24316           | 64824  | 59769  | 109351  | 54      |
+| updateTargetValidatorsLimits                                   | 24317           | 65500  | 62915  | 115676  | 54      |
 ╰----------------------------------------------------------------+-----------------+--------+--------+---------+---------╯
 
 ╭------------------------------------------------+-----------------+--------+--------+--------+---------╮
```

## Checklist

- [x] Appropriate PR labels applied
- [x] Test coverage maintained (`just coverage`)
  - [ ] No need to add/update tests
  - [x] Tests are added/updated
- [x] Documentation maintained
  - [x] No need to update
  - [ ] Updated

dgusakov commented on .github/workflows/test.yml:
These changes can be reverted

dgusakov commented on Justfile:
These as well

madlabman replied:
It still might be needed next time we are out of space.

dgusakov reviewed (commented):
LGTM. Pls revert unnecessary changes about code size limit

madlabman committed:
chore: disable size limit ignore in CI
